Fix greasemonkey window scoping issues

toofar opened the pull request:
Someone reported that OneeChan wasn't working with our greasemonkey implementation, this highlighted an issue with how we were isolating the different scripts, at least compared to other more mature implementations. I tested with Tampermonkey and saw what kind of environment it provided:

1. `window` looks the same as the one the page sees, with any modifications made by the page
2. the global scope looks the same as the page one before the page made any modifications to it, so before javascript started running I suppose
3. saving something to an attribute of `window` make it accessible via the global scope (that is, just referring to it directly)

I tried looking at the TamperMonkey source to see how they did it but I couldn't figure it out. Possibly they just use the environment presented to them by chrome (and get `unsafeWindow` like [this](https://github.com/Tampermonkey/tampermonkey/blob/master/src/content.js#L480))

I am using the ES6 [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) feature to proxy the actual global window and allow for presenting local writes as a mask over top of it. There are some issues around the Proxy implementation does some [seemingly arbitrary type checking](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/handler/get#Invariants) against the target object, designed just to piss me off, that means I had to add another level of closure with another `window` shadow. There were further issues around unbound functions being returned from my `Proxy.get()` implementation, I don't know know why but I added a heuristic to bind those that need it.

I am using the [with](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/with) statement to make that window proxy look like the global scope. Apparently `with` is not cool anymore and if anyone knows how I can achieve this mess without using it that would be great.

PS: I tried again putting the greasemonkey script in a non-main world and that violated point 1 of the above list.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/3793)
<!-- Reviewable:end -->

toofar commented:
Arrghgh :rage4:. This is not working on webkit and the stupid debugger won't let me step through injected JS. I will look at making what the second commit made work on webengine work on webkit too tomorrow, or something.

The-Compiler commented:
Well, that's definitely too crazy JavaScript for me :laughing: Maybe @sandrosc can help with that kind of stuff?

olmokramer commented on qutebrowser/javascript/greasemonkey_wrapper.js:
```diff
@@ -148,9 +148,65 @@
+      // with `with` scope because then it would get assigned to the
+      // proxy and we would get the same problem, so we have to make yet
+      // another nested scope.
+      function qute_gm_window_scope() {  // why can't this be anonymous?
```
This will not work as expected. Take for example this userscript:

```js
var a = 3;
console.log(a, window.a);
```

It logs `3 undefined`, because the `var` declaration is now in function scope, so the variable shall be local to the function. If above code was run in an actual global scope it would print `3 3`.

Even removing the `qute_gm_window_scope` will not work, because `with` is weird, and variable declarations inside the `with` will only forward to the object passed to `with` if the property already exists on the object, and function declarations aren't forwarded whatsoever.

BTW, this function can not be anonymous because function declarations can't be anonymous (how would you call the function afterwards?). You could replace this with

```js
(function() {
    let window = qute_gm_window_proxy;
    // ====== The actual user script source ====== //
    {{ scriptSource }}
    // ====== End User Script ====== //
})();
```

toofar replied:
>It logs `3 undefined`

That isn't unexpected and is exactly what tampermonkey logs, which is what I am copying off. I don't think greasemonkey scripts expect to be in global scope, they just expect the things they write to `window` to appear to be in global scope.

olmokramer replied:
Aha ok then it was just me expecting a global scope. In that case I can't see why the code shouldn't work.

toofar commented:
Yay, I got something working on webkit. Nay, more hacks.
Do you think it is worth adding an option to the `webview` and `webengine` view to enable sending javascript console messages to the python logging facilities? Because getting that working helped me get my test working.

toofar commented:
Nevermind, `Object.assign({}, window).setTimeout(() => {}, 1);` yields `TypeError: Type Error'.

The-Compiler commented:
FWIW If you can't make this work on QtWebKit, I'd also be fine with only doing scope isolation on QtWebEngine (which the majority of people is probably using anyways).

toofar commented:
Yeah, I'm going to think on it a little more though because that would be a pain in itself. I want to do feature detection for when we can't use Proxy but I think the only way to not use the with statement based on a check in greasemonkey_wrapper would be to include the entire source of the greasemonkey script twice, once for the with-Proxy path and one for without, which is yuck. So it'll probably have to be via checking the version in python and using a jinja something.

The-Compiler commented:
Yeah, you should be able to pass `webengine=True` or `webengine=False` to the `template.render` call, and then do something like:

```jinja
{% if webengine %}
    // lots of magic
{{ scriptSource }}
    // lots of magic
{% else %}
{{ scriptSource }}
{% endif %}

toofar commented:
I switched to not doing all the Proxy stuff or any alternative with the current webkit version. I added a new pytest marker to skip the test but left it there for future webkit versions. Hopefully it is sufficient, maybe I should be check for not config.webengine as well.
Because of the feature detection happening in the setup code of the tests I had to turn the setup stuff into a fixture, it seems to work.

The-Compiler commented on qutebrowser/javascript/greasemonkey_wrapper.js:
```diff
@@ -145,9 +145,63 @@
+        },
+        has: function(target, key) {
+          return key in qute_gm_window_shadow ? true :
+                 key in target;
```
This can be simplified as `return key in qute_gm_window_shadow || key in target;`

The-Compiler commented on qutebrowser/javascript/greasemonkey_wrapper.js:
```diff
@@ -145,9 +145,63 @@
+        // more conventionally here. Except we can't do it directly within
+        // with `with` scope because then it would get assigned to the
+        // proxy and we would get the same problem, so we have to make yet
+        // another nested scope.
```
I don't get why the function-IIFE is needed. Shouldn't `let` take care of making it block-scoped (with the userscript) instead of assigning it to the proxy?

However, the tests also don't seem to catch this - if I remove the function, they don't fail. If I also remove the `let`, they still don't fail :wink:

toofar replied:
Thank you for being so thorough, I guess I was doing something silly (not using a keyword?) and then did that to code around that and then later stopped doing the first silly thing. Everything appears to be fine without it, I also changed the `let` to a `const` because why not?

The-Compiler commented on qutebrowser/javascript/greasemonkey_wrapper.js:
```diff
@@ -145,9 +145,63 @@
+        // another nested scope.
+        (function() {
+          let window = qute_gm_window_proxy;
+          // ====== The actual user script source ====== //
```
Shouldn't those comments be repeated around the second `{{ scriptSource }}` too, so someone looking at the processed JS source on QtWebKit also sees them?

toofar commented on qutebrowser/browser/greasemonkey.py:
```diff
@@ -108,13 +109,18 @@ def code(self):
+        # Don't use Proxy on this webkit version, the support isn't there.
+        use_proxy = not (
+            objects.backend == usertypes.Backend.QtWebKit and
+            version.qWebKitVersion() == '602.1')
```
This is failing under test in an environment that doesn't have webkit installed because `test/conftest.py::set_backend()` is setting `objects.backend` to `usertypes.Backend.QtWebKit` by default so the second clause is being executed but `version.qWebKitVersion` is `None`. Should I change it like so?
```diff
diff --git i/tests/conftest.py w/tests/conftest.py
index 97d5623372a9..5043abb4dd5d 100644
--- i/tests/conftest.py
+++ w/tests/conftest.py
@@ -223,8 +223,10 @@ def check_display(request):
 @pytest.fixture(autouse=True)
 def set_backend(monkeypatch, request):
     """Make sure the backend global is set."""
-    backend = (usertypes.Backend.QtWebEngine if request.config.webengine
-               else usertypes.Backend.QtWebKit)
+    if not request.config.webengine and version.qWebKitVersion:
+        backend = usertypes.Backend.QtWebKit
+    else:
+        backend = usertypes.Backend.QtWebEngine
     monkeypatch.setattr(objects, 'backend', backend)
 
```

toofar commented:
Okay, I squashed/rebase the two smaller things, I left the removing IIFE in a separate commit because I also changed the block scoped window from `let` to `const`.
I also added a commit that changes a global autouse fixture because it was breaking some tests in a chroot where the new qtwebkit wasn't available.

The-Compiler commented:
Looks great now - thanks a lot! :+1:
